### ISSUE TITLE
Roslyn seems to be substantially slower in the latest version

### ISSUE THREAD

**Issue by jroessel** [dotnet/roslyn#35992]
**Version Used**: 3.1.0

We have an internal compiler from C# to other languages based on Roslyn. The general approach is walking the syntax trees and producing source code that looks mostly similar to the original source. Runtime seems to be mostly dominated by Roslyn in looking up symbols for tokens (but it's hard to profile a complete 13-minute compilation).

I recently upgraded to the latest version of Microsoft.CodeAnalysis (3.1.0) from the version we've used so far (1.3.2) and noticed a slowdown by about 50 %. I can't really point to a specific place, as I haven't profiled a complete run with either version, but downgrading to 2.10 resolved the problem again.

**Comment by jmarolf** [comment]
@jroessel could you record a perf trace in [perfview](https://github.com/dotnet/roslyn/wiki/Recording-performance-traces-with-PerfView) for both 2.10 and 3.1.0 versions of the compiler or point us at your source?

**Comment by jroessel** [comment]
Uploaded two profiles [https://files.yworks.com/share/rHj9db5r](here). The link should work for 15 days. Our source isn't public, as it's an internal tool.

The run is for a small codebase we use for testing, so it took less time (but the ETL file is still quite large). Our source code is mostly identical between both runs except for the Roslyn dependency, of course. With 2.10 it took 14 seconds, with 3.1.0 it took 48 seconds. For other (larger) conversions I've seen changes from 11 to 17 minutes.

Most of the time is usually spent getting symbols for tokens/expressions/etc. since, well, for a conversion we have to know what we're converting, so that is done for pretty much every span in the source and in total that takes a while.

**Comment by sharwell** [comment]
@mavasani Is there an API to get a fully-bound semantic model so calls to `GetSymbolInfo` and `GetTypeInfo` don't have to recompute everything?

**Comment by jmarolf** [comment]
@jroessel thanks! For internal MS employees I've uploaded the traces here: \\mlangfs1\public\jmarolf\issue_35992_traces

**Comment by jmarolf** [comment]
Looking at the trace, binding is simply taking longer in 3.1.0


Name | 2.10.0 | 3.1.0| Delta | Responsibility % | Overweight % | Interest Level
-- | -- | -- | -- | -- | -- | --
microsoft.codeanalysis.csharp!Microsoft.CodeAnalysis.CSharp.LocalBinderFactory.VisitMethodDeclaration(class  Microsoft.CodeAnalysis.CSharp.Syntax.MethodDeclarationSyntax) | 4.0 | 15895.0 | 15891.0 | 44.64 | 179050.50 | 5
microsoft.codeanalysis!Microsoft.CodeAnalysis.ChildSyntaxList.ItemInternal(class  Microsoft.CodeAnalysis.SyntaxNode,int32) | 138.0 | 11875.0 | 11737.0 | 32.97 | 3833.21 | 5
microsoft.codeanalysis.csharp!Microsoft.CodeAnalysis.CSharp.LocalScopeBinder.get_Locals() | 519.0 | 17433.0 | 16914.0 | 47.52 | 1468.80 | 5
microsoft.codeanalysis.csharp!Microsoft.CodeAnalysis.CSharp.LocalBinderFactory.VisitBlock(class  Microsoft.CodeAnalysis.CSharp.Syntax.BlockSyntax) | 501.0 | 15897.0 | 15396.0 | 43.25 | 1385.02 | 5
microsoft.codeanalysis.csharp!Microsoft.CodeAnalysis.CSharp.LocalScopeBinder.get_LocalsMap() | 553.0 | 17457.0 | 16904.0 | 47.49 | 1377.68 | 5
microsoft.codeanalysis.csharp!Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetEnclosingBinderInternal(int32) | 951.0 | 16462.0 | 15511.0 | 43.58 | 735.10 | 5
microsoft.codeanalysis.csharp!Microsoft.CodeAnalysis.CSharp.Binder.BindLeftOfPotentialColorColorMemberAccess(class  Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax,class  Microsoft.CodeAnalysis.DiagnosticBag) | 1077.0 | 18553.0 | 17476.0 | 49.10 | 731.33 | 5

**Comment by jroessel** [comment]
@sharwell So currently Roslyn has to recompute everything when querying anything on a Semantic model? In that case I'm not too surprised that csc can compile the same codebase in less than a minute while we need over 10 minutes.

Oh my ... I just made a small change to fetch the model only once per tree and got a speedup of 2–3×. Grabbing the model anew every single time was something we've done for a long time and never really questioned it (I faintly remember Roslyn caching many things internally to only recompute when things change, so I may have assumed this to just work, despite us fetching a new model every time).

I guess it _is_ intended to grab the model once, perform all kinds of semantic questionings on it and only throw it away when no longer needed?

**Comment by sharwell** [comment]
@jroessel I'm not exactly sure what the conditions for recomputing are, but it sounds from your statement that one of them involves discarding data in the semantic model when a new one is created.

**Comment by mavasani** [comment]
> Grabbing the model anew every single time was something we've done for a long time and never really questioned it 

[Compilation.GetSemanticModel](http://source.roslyn.io/#Microsoft.CodeAnalysis/Compilation/Compilation.cs,198) API always creates a [completely new semantic model](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Compilation/CSharpCompilation.cs,1932), as also mentioned in the doc comments of the APIs. Any semantic query on this new semantic model will bind almost everything from scratch, but might be able to re-use some computed data from internal per-symbol caches. It always recommended that clients re-use semantic model across semantic queries for the same tree to avoid additional binding costs.

**Comment by jmarolf** [comment]
So, getting a semantic model should be in-expensive but asking questions of it may not be.  If you modify a syntax tree, get a new semantic model for that modified tree, and ask a question we will need to rebind everything in order to answer it. As @mavasani 	says, any re-use that happens is an implementation detail that you can't rely on. There will always be a worst case where we can't share anything.

**Comment by jroessel** [comment]
Well, learned something new today (and revisited code I've written in a prototype stage four years ago). I'll redo the profiles, in that case, as my usage was not typical and was expected to be slow to begin with. 

Indeed, after profiling the times are pretty much the same, so I guess this issue has just been my own stupidity.

**Comment by sharwell** [comment]
> ... has just been my own stupidity and can be resolved ...

It wasn't obvious to me either at first, and I work on this project full time. 😄 

I am glad that we were able to help get this solved for you so quickly though!

**Comment by SomeoneIsWorking** [comment]
Now it takes like a minute in VSCode when I want to refactor something using Roslyn.
I move to the part with the suggestion. I press CMD+dot, it takes forever for the suggestions to appear.
